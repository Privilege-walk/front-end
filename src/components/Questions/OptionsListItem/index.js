import React, { useState } from "react";
import {Input} from 'reactstrap';
import Button from '@mui/material/Button';

import Box from '@mui/material/Box';
import FormControl from '@mui/material/FormControl';


export default function OptionsListItem({ id, type, description, value, addOption, editOption, deleteOption, optionDisabled }) {
    const [optionDesc, setOptionDesc] = useState(description);
    const [optionPoints, setOptionPoints] = useState(value);
    const [isDisabled, setIsDisabled] = useState(optionDisabled);

    function validateOption() {
        let optiondesc = optionDesc === "" || optionDesc.trim() === "";
        let optionvalue = optionPoints === "" || optionPoints.trim() === "";
        return optiondesc || optionvalue;
    }

    function handleSubmit() {
        if (type === 'new') {
            addOption({
                id: id,
                description: optionDesc,
                value: optionPoints
            });
            setOptionDesc('');
            setOptionPoints('');
        } else {
            editOption({
                id: id,
                description: optionDesc,
                value: optionPoints
            });
            setIsDisabled(true);
        }
    }

    return (
        <div style={{p: '2px 4px', display: 'flex', flexGrow: 1, alignItems: 'center'}}>
            <FormControl sx={{ m: 1, width: '350px' }}>
                <Input
                    id={"new-option-desc-"+id}
                    data-testid={"new-option-desc-"+id}
                    placeholder="Option Description"
                    value={optionDesc}
                    onChange={(e) => setOptionDesc(e.target.value)}
                    disabled={isDisabled}
                />
            </FormControl>
            <FormControl sx={{ m: 1, width: '125px' }}>
                <Input
                    id={"new-option-points-"+id}
                    data-testid={"new-option-points-"+id}
                    placeholder="Option Points"
                    value={optionPoints}
                    onChange={(e) => setOptionPoints(e.target.value)}
                    disabled={isDisabled}
                />
            </FormControl>
            {
                (type === 'new' || (type === 'edit' && !isDisabled)) && 
                (<Box sx={{p: '0px 0px 0px 8px'}}>
                    <Button 
                        id={"add-"+id}
                        data-testid={"add-"+id}
                        variant="contained"
                        type="submit"
                        className="ml-2"
                        onClick={handleSubmit}
                        disabled={validateOption()}
                    >
                        {type === 'new' ? 'Add' : 'Submit'}
                    </Button>  
                </Box>)
            }
            {
                type === 'edit' && isDisabled && 
                (<Box sx={{p: '0px 0px 0px 8px'}}>
                    <Button 
                        id={"edit-"+id}
                        data-testid={"edit-"+id}
                        variant="contained"
                        type="submit"
                        className="ml-2"
                        onClick={() => setIsDisabled(false)}
                    >
                        Edit
                    </Button>  
                </Box>)
            }
            {
                type === 'edit' && isDisabled && 
                (<Box sx={{p: '0px 0px 0px 8px'}}>
                    <Button 
                        id={"delete-"+id}
                        data-testid={"delete-"+id}
                        variant="contained"
                        type="submit"
                        className="ml-2"
                        onClick={() => deleteOption(id)}
                    >
                        Delete
                    </Button>  
                </Box>)
            }
        </div>
    );
}