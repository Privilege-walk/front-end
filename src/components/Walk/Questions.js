import React from "react";
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import Container from '@mui/material/Container';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';

import CardContent from '@mui/material/CardContent';


export default class Questions extends React.Component{
    
    constructor(props){
        super(props);
    }
    
    handleSelectChoice(choiceId, selected){
        let answer = selected? "" : choiceId;
        const props = this.props;
        let answers = { ...props.answers };
        let questionId = props.questions[props.questionIndex].id;
        answers[questionId] = answer;
        props.setAnswers(answers);
    }

    renderHeader(){
        return (
            <Grid container lg={1} item direction='column'>
                {/* Title */}
                <Grid item>
                    <Typography id="eventName"  variant="h5" component="div">
                        Event: {this.props.eventName}
                    </Typography>
                </Grid>

            {/* Graph */}
            {/* <Grid sx={{border: '2px solid'}} item>
                <Typography  variant="h5" component="div">
                </Typography>
            </Grid> */}
        </Grid>
        );
    }

    renderBody(){
        let question = ""; 
        let choices = [];
        let questionDetails = this.props.questions[this.props.questionIndex];
        if(questionDetails){
            question = questionDetails.description;
            choices = questionDetails.choices;
        }
        return (
            <Grid container lg={10} sx={{ minHeight: '225px'}} item direction='column'>
                {/* Question */}
                <Grid item>
                    <Typography id={"question"}  variant="h6" component="div">
                        {questionDetails? 
                            questionDetails.description : 
                            ""
                        }
                    </Typography>
                </Grid>

                {/* Choices */}
                <Grid 
                    sx={{mb: 1}} 
                    id="choices"
                    container item 
                    alignItems="flex-start" 
                    direction="column" 
                >
                    {choices.map((choice, _) => (
                        choice.id == this.props.answers[questionDetails.id]?
                            (
                                <Button
                                    variant="outlined" 
                                    onClick={() => this.handleSelectChoice(choice.id, true)}
                                    sx={{mx:1, mt:1}} 
                                    className="choice"
                                    style={{ border: '3px solid' }}
                                    key={choice.id}
                                >
                                    {choice.description}
                                </Button>
                            ):
                            (
                                <Button 
                                    variant="outlined" 
                                    className="choice"
                                    onClick={() => this.handleSelectChoice(choice.id, false)}
                                    sx={{mx:1, mt:1}} 
                                    key={choice.id}
                                >
                                    {choice.description}
                                </Button>
                            )
                        )
                    )}
                </Grid>
            </Grid>
        );
    }

    renderHostBody(){
        let question = ""; 
        let questionDetails = this.props.questions[this.props.questionIndex];
        if(questionDetails){
            question = questionDetails.description;
        }
        return (
            <Grid container lg={10} sx={{ minHeight: '225px'}} item direction='column'>
                {/* Question */}
                <Grid item>
                    <Typography id={"question"}  variant="h6" component="div">
                        {questionDetails? 
                            questionDetails.description : 
                            ""
                        }
                    </Typography>
                </Grid>
                <Grid container item>
                    <Card variant="outlined">
                    <CardContent>
                        <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>
                            Total Participants
                        </Typography>
                        <Typography variant="h5" component="div">
                            {this.props.activeUsers}
                        </Typography>       
                    </CardContent>  
                    </Card>

                    <Card variant="outlined">
                    <CardContent>
                        <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>
                            Total Answered
                        </Typography>
                        <Typography variant="h5" component="div">
                            {this.props.answeredUsers}
                        </Typography>       
                    </CardContent> 
                    </Card> 
                </Grid>
            </Grid>);
    }

    renderFooter(){
        const question = this.props.questions[this.props.questionIndex];
        const questionIndex = question? question.id: "";
        return (
            <Grid container lg={1} sx={{minHeight: '50px'}} item direction='column'>
                {/* Number of people who have answered so far */}
                {/* <Grid item>
                    <Typography  variant="h6" component="div"> 
                    </Typography>
                </Grid> */}

                <Grid container direction="column" item >

                    <Typography id="numQuestions"  variant="p" component="div">
                        {this.props.questionIndex + 1}/{this.props.questions.length} Questions
                    </Typography>
                    
                </Grid>

                {this.props.userType == "HOST"?
                    (
                        <Button 
                            variant={this.props.activeUsers == this.props.answeredUsers? "contained" : "outlined"}
                            type="submit"
                            sx={{mt:2}}
                            onClick={()=> this.props.nextQuestion()}                         
                        >
                            Next Question
                        </Button>
                    ) :
                    (
                        <Button 
                            variant="outlined"
                            type="submit"
                            sx={{mt:2}}
                            onClick={()=> {}}     
                            disabled={!this.props.answers[questionIndex]}        
                        >
                            
                            Waiting for host ...
                        </Button>
                    )
                        }  
            </Grid>
        );
    }

    render(){
        switch(this.props.userType){
            case 'HOST':
                return (
                    <Container 
                        sx={{px:6, mt:3}}
                    >
                        <Grid 
                            sx={{minHeight: '40px'}} 
                            container 
                            justifyContent='center' 
                            alignItems='center' 
                            direction='column' 
                            spacing={2}
                        ></Grid>
                        {this.renderHeader()}
                        {this.renderHostBody()}
                        {this.renderFooter()}
                    </Container>
                )
        }
        return (        
            <Container 
                sx={{px:6, mt:3}}
            >
                <Grid 
                    sx={{minHeight: '40px'}} 
                    container 
                    justifyContent='center' 
                    alignItems='center' 
                    direction='column' 
                    spacing={2}
                >
                    {/* Header */}
                    {this.renderHeader()}

                    {/* Body */}
                    {this.renderBody()}

                    {/* Footer */}
                    {this.renderFooter()}
                    
                </Grid>
            </Container>
        );
    }
}