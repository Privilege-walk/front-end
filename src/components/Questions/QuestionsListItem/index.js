import React, { useState } from "react";
 import Card from '@mui/material/Card';
 import CardContent from '@mui/material/CardContent';
 import Typography from '@mui/material/Typography';
 import OptionsListItem from "../OptionsListItem";


export default function QuestionsListItem({ description, choices }) {
    const [answerChoices, changeChoices] = useState([...choices].reverse())
    return (
        <Card className="">
            <CardContent>
            <Typography>{description}</Typography>
            </CardContent>
            {
                answerChoices.map( (item) => (
                    <OptionsListItem 
                        key={item.id}
                        type = 'view'
                        description = {item.description}
                        value = {item.value}
                        optionDisabled = {true}
                    />
                ))
            }
        </Card>
    );
}