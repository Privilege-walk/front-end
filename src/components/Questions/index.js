import React, { useState, useEffect } from "react";
import { connect } from 'react-redux';
import { useLocation } from "react-router-dom";
import {Row, Container} from 'reactstrap';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Button from '@mui/material/Button';
import InputLabel from '@mui/material/InputLabel';
import TableHead from '@mui/material/TableHead';

import Box from '@mui/material/Box';
import FormControl from '@mui/material/FormControl';
import OutlinedInput from '@mui/material/OutlinedInput';
import QuestionsListItem from "./QuestionsListItem";
import OptionsListItem from "./OptionsListItem";
import { fetchQuestions, createQuestion } from '../../Store/actions';


function Questions({fetchQuestions, createQuestion}) {
    const props = useLocation();
    const [questionsList, setQuestionsList] = useState([]);
    const [optionsList, setOptionsList] = useState([]);
    const [newOptionId, setNewOptionId] = useState(0);
    const [newQuestion, setNewQuestion] = useState("");
    const [errMsg, setErrMsg] = useState("");

    useEffect(async () => fetchAllQuestions(), []);

    async function fetchAllQuestions() {
        const action = await fetchQuestions({ params: {event_id: props.state.id} });
        if (action.payload.error){
            setErrMsg("Unable to display the questions list!");
        }else{
            setQuestionsList(action.payload.questions);
        }
    }
    
    function createChoices() {
        let options = JSON.parse(JSON.stringify(optionsList));
        return options.map( item => {
            return { "description": item.description, "value": item.value };
        });
    }

    async function submitNewQuestion(event) {
        let requestBody = {
            "event_id": props.state.id,
            "title": newQuestion,
            "choices": createChoices()
        }
        const action = await createQuestion({ requestBody });
        if (action.payload.status == "created"){
            fetchAllQuestions();
            setNewQuestion(action.payload.newQuestion);
            setOptionsList(action.payload.optionsList);
            setNewOptionId(action.payload.newOptionId);
        }else{
            setErrMsg("Unable to create the question!");
        }
    }

    function handleOptionAdd(option) {
        let options = JSON.parse(JSON.stringify(optionsList));
        options.push(option);
        setOptionsList(options);
        setNewOptionId(newOptionId + 1);
    }

    function handleOptionEdit(option) {
        let options = JSON.parse(JSON.stringify(optionsList));
        options.forEach( (item, index) => {
            if(item.id === option.id) {
                options[index] = option;
            }
        });
        setOptionsList(options);
    }

    function handleOptionDelete(id) {
        let options = JSON.parse(JSON.stringify(optionsList));
        options = options.filter(x => x.id != id);
        setOptionsList(options);
    }

    return (
        <Container className="mx-2">
            <Row>
                <h1 className="event">{props.state.name}</h1>
                <Paper 
                    className="mb-3 align-items-center" 
                    variant="outlined" 
                    elevation={0} 
                    sx={{ p: '2px 4px', display: 'flex', flexGrow: 1, alignItems: 'center', flexDirection: 'column'}}
                >
                    <FormControl fullWidth sx={{ m: 1 }}>
                        <InputLabel htmlFor="new-question-name">Add New Question</InputLabel>
                        <OutlinedInput
                            id="new-question-name"
                            label="Add New Question"
                            variant="filled"
                            autoComplete="off"
                            value={newQuestion}
                            onChange={(e) => setNewQuestion(e.target.value)}
                        />
                    </FormControl>
                    <div style={{width: '100%'}}>
                        {
                            optionsList.map( (item) => (
                                <OptionsListItem 
                                    key={item.id}
                                    id = {item.id}
                                    type = 'edit'
                                    description = {item.description}
                                    value = {item.value}
                                    editOption = {handleOptionEdit}
                                    deleteOption = {handleOptionDelete}
                                    optionDisabled = {true}
                                />
                            ))
                        }
                        <OptionsListItem 
                            id = {newOptionId}
                            type = 'new'
                            description = ''
                            value = ''
                            addOption = {handleOptionAdd}
                            optionDisabled = {false}
                        />
                        <Box sx={{p: '20px 0px 8px 8px'}}>
                            <Button 
                                variant="contained"
                                type="submit"
                                className="ml-2"
                                onClick={submitNewQuestion}
                                disabled={(newQuestion === "" || newQuestion.trim() === "") || optionsList.length < 2}
                            >
                                Create Question
                            </Button>  
                        </Box>
                    </div>
                </Paper>
            </Row>

            <TableContainer component={Paper}>
                <Table sx={{ minWidth: 650 }} aria-label="simple table">
                    <TableHead>
                        <TableRow>
                            <TableCell>
                                <h4>Questions</h4>
                            </TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                    {
                        questionsList.length === 0 ?
                        (<TableRow style={{color:'gray',marginLeft:10}}>
                            <TableCell><i>No questions to show.</i></TableCell>
                        </TableRow>)
                        :
                        (
                            questionsList.map( (item, _) => (
                                <TableRow key={item.id}>
                                    <TableCell>
                                        <QuestionsListItem 
                                            key={item.id}
                                            description={item.description}
                                            choices={item.choices}
                                        />
                                    </TableCell>
                                </TableRow>)
                        ))
                    }
                    </TableBody>
                </Table>
            </TableContainer>
        </Container>
    );
}

const mapStateToProps = state => {
    return {
        token: state.authToken
    };
};

const mapDispatchToProps = {
    fetchQuestions,
    createQuestion
};

var QuestionsContainer = connect(mapStateToProps, mapDispatchToProps)(Questions);
export default QuestionsContainer;