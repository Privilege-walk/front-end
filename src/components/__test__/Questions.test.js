import React from "react";
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Questions from '../Questions';

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const mockCreateQuestion = jest.fn();
const mockFetchQuestions = jest.fn();
const mockedUsedLocation = jest.fn();
jest.mock('../../Store/actions', () => ({
    createQuestion: () => mockCreateQuestion,
    fetchQuestions: () => mockFetchQuestions
}));
jest.mock("react-router-dom", () => ({
    Navigate: () => "Navigate",
    useLocation: () => mockedUsedLocation(),
}))


// Snapshot test
const setup = (numQuestions=0) => {
    let utils;
    const store = mockStore({
        token: "token"
    });
    let questions = [];
    for (let i = 0; i < numQuestions; i++){
        questions.push({
            id: i,
            description: "Question " + i,
            choices: [
                { id: 876, description: "Pizza", value: 1 },
                { id: 54, description: "Ice Cream", value: 2 }
            ]
        });
    }
    mockFetchQuestions.mockReturnValue({payload: {questions}});
    mockCreateQuestion.mockReturnValue({
        payload: {status: "created", newQuestion: "", optionsList: [], newOptionId: 0}
    });
    mockedUsedLocation.mockReturnValue({state: {id: 1, name: "New Event"}});
    act (() => {
        utils = render(
            <Provider store={store}>
                <Questions />
            </Provider>
        );
    });

    const addQuestionInput = screen.getByLabelText('Add New Question');
    const addOptionInput = screen.getByPlaceholderText("Option Description");
    const addPointsInput = screen.getByPlaceholderText("Option Points");
    const addOptionBtn = screen.getByText("Add", {selector: "button"});
    const addQuestionBtn = screen.getByText("Create Question", {selector: "button"});

    return {
        addQuestionInput,
        addOptionInput,
        addPointsInput,
        addOptionBtn,
        addQuestionBtn,
        ...utils
    }
}

test("renders Question page without crashing", () => {
    let utils = setup();
    const { asFragment } = utils;
    expect(asFragment()).toMatchSnapshot();
});

test("create question successfully", async () => {
    let utils = setup(1);
    const { 
        addQuestionInput,
        addQuestionBtn,
    } = utils;
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', true);

    fireEvent.change(addQuestionInput, {target: {value: "Question 1?"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    fireEvent.change(screen.getByTestId('new-option-desc-0'), {target: {value: "option 1"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-0'), {target: {value: "+1"}});
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-0'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', true);

    fireEvent.change(screen.getByTestId('new-option-desc-1'), {target: {value: "option 2"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-1'), {target: {value: "-1"}});
    expect(screen.getByTestId('add-1')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-1'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', false);
    
    await act(async () => {
        await fireEvent.click(addQuestionBtn);
    });

    expect(addQuestionInput.value).toEqual("");
    expect(screen.getByTestId('new-option-desc-0').value).toEqual("");
    expect(screen.getByTestId('new-option-points-0').value).toEqual("");
})

test("check edit and delete options", async () => {
    let utils = setup();
    const { 
        addQuestionInput,
        addQuestionBtn,
    } = utils;
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', true);

    fireEvent.change(addQuestionInput, {target: {value: "Question 1?"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    fireEvent.change(screen.getByTestId('new-option-desc-0'), {target: {value: "option 1"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-0'), {target: {value: "+1"}});
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-0'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', true);

    fireEvent.change(screen.getByTestId('new-option-desc-1'), {target: {value: "option 2"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-1'), {target: {value: "-1"}});
    expect(screen.getByTestId('add-1')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-1'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', false);

    await act(async () => {   
        await fireEvent.click(screen.getByTestId('delete-1'));
    });
    expect(addQuestionBtn).toHaveProperty('disabled', true);

    expect(screen.getByTestId('new-option-desc-0').disabled).toEqual(true);
    await act(async () => {
        await fireEvent.click(screen.getByTestId('edit-0'));
    });
    expect(screen.getByTestId('new-option-desc-0').disabled).toEqual(false);

    await act(async () => {
        await fireEvent.click(screen.getByTestId('add-0'));
    });
    expect(screen.getByTestId('new-option-desc-0').disabled).toEqual(true);
})

test("create question failed", async () => {
    let utils = setup();
    const { 
        addQuestionInput,
        addQuestionBtn,
    } = utils;
    mockCreateQuestion.mockReturnValue({
        payload: {errors: "error"}
    });
    mockFetchQuestions.mockReturnValue({payload: {questions:[]}});

    expect(addQuestionBtn).toHaveProperty('disabled', true);
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', true);

    fireEvent.change(addQuestionInput, {target: {value: "Question 1?"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    fireEvent.change(screen.getByTestId('new-option-desc-0'), {target: {value: "option 1"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-0'), {target: {value: "+1"}});
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-0'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', true);

    fireEvent.change(screen.getByTestId('new-option-desc-1'), {target: {value: "option 2"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-1'), {target: {value: "-1"}});
    expect(screen.getByTestId('add-1')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-1'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', false);
    
    await act(async () => {
        await fireEvent.click(addQuestionBtn);
    });
    expect(addQuestionInput.value).not.toEqual("");
})

test("fetch questions empty list", async () => {
    let utils = setup();
    const { 
        addQuestionInput,
        addQuestionBtn,
    } = utils;
    mockCreateQuestion.mockReturnValue({
        payload: {errors: "error"}
    });
    mockFetchQuestions.mockReturnValue({payload: {questions:[]}});

    expect(addQuestionBtn).toHaveProperty('disabled', true);
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', true);

    fireEvent.change(addQuestionInput, {target: {value: "Question 1?"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    fireEvent.change(screen.getByTestId('new-option-desc-0'), {target: {value: "option 1"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-0'), {target: {value: "+1"}});
    expect(screen.getByTestId('add-0')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-0'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', true);

    fireEvent.change(screen.getByTestId('new-option-desc-1'), {target: {value: "option 2"}});
    expect(addQuestionBtn).toHaveProperty('disabled', true);
    
    fireEvent.change(screen.getByTestId('new-option-points-1'), {target: {value: "-1"}});
    expect(screen.getByTestId('add-1')).toHaveProperty('disabled', false);
    await act(async () => {   
        await fireEvent.click(screen.getByTestId('add-1'));
    });

    expect(addQuestionBtn).toHaveProperty('disabled', false);
    
    await act(async () => {
        await fireEvent.click(addQuestionBtn);
    });
    expect(addQuestionInput.value).not.toEqual("");
})