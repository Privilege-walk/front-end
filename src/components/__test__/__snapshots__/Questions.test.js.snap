// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Question page without crashing 1`] = `
<DocumentFragment>
  <div
    class="mx-2 container"
  >
    <div
      class="row"
    >
      <h1
        class="event"
      >
        New Event
      </h1>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded mb-3 align-items-center css-101q0pl-MuiPaper-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-1w4vsez-MuiFormControl-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="new-question-name"
          >
            Add New Question
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            variant="filled"
          >
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="new-question-name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Add New Question
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          style="width: 100%;"
        >
          <div
            style="display: flex; flex-grow: 1; align-items: center;"
          >
            <div
              class="MuiFormControl-root css-104sb8r-MuiFormControl-root"
            >
              <input
                class="form-control"
                data-testid="new-option-desc-0"
                id="new-option-desc-0"
                placeholder="Option Description"
                type="text"
                value=""
              />
            </div>
            <div
              class="MuiFormControl-root css-3t6boo-MuiFormControl-root"
            >
              <input
                class="form-control"
                data-testid="new-option-points-0"
                id="new-option-points-0"
                placeholder="Option Points"
                type="text"
                value=""
              />
            </div>
            <div
              class="MuiBox-root css-1ldu1i8"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled ml-2 css-sghohy-MuiButtonBase-root-MuiButton-root"
                data-testid="add-0"
                disabled=""
                id="add-0"
                tabindex="-1"
                type="submit"
              >
                Add
              </button>
            </div>
          </div>
          <div
            class="MuiBox-root css-k1clu9"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled ml-2 css-sghohy-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              Create Question
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              <h4>
                Questions
              </h4>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            style="color: gray; margin-left: 10px;"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <i>
                No questions to show.
              </i>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
